src/raftstore/store/snap.rs

```diff
@@ -618,6 +665,987 @@ mod v1 {
+                              self.path(),
+                              e);
+                        self.delete();
+                        try!(self.init_for_building(snap));
+                    }
+                }
+            }
+
+            let mut snap_key_count = 0;
+            let (begin_key, end_key) = (enc_start_key(region), enc_end_key(region));
+            for cf in SNAPSHOT_CFS {
+                try!(self.switch_to_cf_file(cf));
+                let (cf_key_count, cf_size) = if plain_file_used(cf) {
+                    let file = self.cf_files[self.cf_index].file.as_mut().unwrap();
+                    try!(build_plain_cf_file(file, snap, cf, &begin_key, &end_key))
+                } else {
+                    let mut key_count = 0;
+                    let mut size = 0;
+                    try!(snap.scan_cf(cf,
+                                      &begin_key,
+                                      &end_key,
+                                      false,
+                                      &mut |key, value| {
+                                          key_count += 1;
+                                          size += key.len() + value.len();
+                                          try!(self.add_kv(key, value));
+                                          Ok(true)
+                                      }));
+                    (key_count, size)
+                };
+                snap_key_count += cf_key_count;
+                SNAPSHOT_CF_KV_COUNT.with_label_values(&[cf]).observe(cf_key_count as f64);
+                SNAPSHOT_CF_SIZE.with_label_values(&[cf]).observe(cf_size as f64);
+                info!("[region {}] scan snapshot {}, cf {}, key count {}, size {}",
+                      region.get_id(),
+                      self.path(),
+                      cf,
+                      cf_key_count,
+                      cf_size);
+            }
+
+            try!(self.save_cf_files());
+            stat.kv_count = snap_key_count;
+            // save snapshot meta to meta file
+            let snapshot_meta = try!(get_snapshot_meta(&self.cf_files[..]));
+            self.meta_file.meta = snapshot_meta;
+            try!(self.save_meta_file());
+
+            Ok(())
+        }
+    }
+
+    impl Snapshot for Snap {
+        fn build(&mut self,
+                 snap: &DbSnapshot,
+                 region: &Region,
+                 snap_data: &mut RaftSnapshotData,
+                 stat: &mut SnapshotStatistics)
+                 -> RaftStoreResult<()> {
+            let t = Instant::now();
+            try!(self.do_build(snap, region, stat));
+
+            // set snapshot meta data
+            let total_size = try!(self.total_size());
+            snap_data.set_file_size(total_size);
+            snap_data.set_version(SNAPSHOT_VERSION);
+            snap_data.set_meta(self.meta_file.meta.clone());
+
+            // add size
+            let mut size_track = self.size_track.wl();
```

Comment:
The size should be updated once the snapshot file is saved, otherwise if there are several failure, there will be a big gap between the size_track and the actual snap directory size.